runtime/lua/vim/shared.lua

```diff
@@ -1151,34 +1154,70 @@ end
 --- @field keeppatterns? boolean
 --- @field lockmarks? boolean
 --- @field noautocmd? boolean
---- @field options? table<string, any>
+--- @field o? table<string, any>
 --- @field sandbox? boolean
 --- @field silent? boolean
 --- @field unsilent? boolean
 --- @field win? integer
+--- @field wo? table<string, any>
+
+local save_set_context = function(context)
```

Comment:
The name might need a bit of explanation, is this correct:
```suggestion
--- Gets option :set context.
local save_set_context = function(context)
```

Nit: "save" implies side-effects, so "get" would normally be preferrable, but `get_set_context` would be strange.

Reply:
> How about `cache_and_set_context()`?

doesn't seem better than "save"

Reply:
> How about `cache_and_set_context()`?

doesn't seem better than "save".  I see now that this is setting options so it does in fact have side effects, so ignore my previous comment. Just maybe a brief docstring would help.

Reply:
I resolved this by compactifying implementation so that it can be reasonably fit inside `vim._with()` body. No helper function - no problem.

Reply:
I've recently had a realization that if there is a misplaced option scope (like `{ bo = { winblend = 10 } }`), then the approach of "cache and set" inside same double loop would not properly restore already set context values (due to error during `vim.bo.winblend = 10`). So I indeed had to separate `get_context_state()` into a separate function (with brief description of what it does).